stdlib/asm/word.masm

```diff
@@ -0,0 +1,222 @@
+
+        movdn.3
+        # => [2_x, 1_x, is_lhs_greater, is_felt_eq, continue, <remaining_felts>]
+
+        # check 1_x > 2_x
+        gt
+        # => [is_felt_gt, is_lhs_greater, is_felt_eq, continue, <remaining_felts>]
+
+        dup.3 and
+        # => [is_felt_gt_if_continue, is_lhs_greater, is_felt_eq, continue, <remaining_felts>]
+
+        or movdn.2
+        # => [is_felt_eq, continue, is_lhs_greater, <remaining_felts>]
+
+        # keeps continue at 1 if the felts are equal
+        # sets continue to 0 if the felts are not equal
+        and
+        # => [continue, is_lhs_greater, <remaining_felts>]
+
+        swap
+        # => [is_lhs_greater, continue, <remaining_felts>]
+    end
+    # => [is_lhs_greater, continue]
+
+    swap drop
+    # => [is_lhs_greater]
+end
+
+#! Returns true if LHS is greater than or equal to RHS.
+#!
+#! Inputs:  [LHS, RHS]
+#! Outputs: [is_lhs_greater_or_equal]
+#!
+#! Cycles: 118
+export.gte
+    exec.lt
+    not
+end
+
+#! Returns true if LHS is strictly less than RHS, false otherwise.
+#!
+#! The implementation avoids branching for performance reasons.
+#!
+#! From an implementation standpoint this is exactly the same as `word::gt` except it uses
+#! `lt` rather than `gt`. See its docs for details.
+#!
+#! Inputs:  [LHS, RHS]
+#! Outputs: [is_lhs_lesser]
+#!
+#! Cycles: 117
+export.lt
+    exec.arrange_words_adjacent
+    # => [2_3, 1_3, 2_2, 1_2, 2_1, 1_1, 2_0, 1_0]
+
+    push.1.0
+    # => [is_lhs_less, continue, 2_3, 1_3, 2_2, 1_2, 2_1, 1_1, 2_0, 1_0]
+
+    repeat.4
+        movup.3 movup.3
+        # => [2_x, 1_x, is_lhs_less, continue, <remaining_felts>]
+
+        # check 1_x == 2_x; if so, we continue
+        dup dup.2 eq
+        # => [is_felt_eq, 2_x, 1_x, is_lhs_less, continue, <remaining_felts>]
+
+        movdn.3
+        # => [2_x, 1_x, is_lhs_less, is_felt_eq, continue, <remaining_felts>]
+
+        # check 1_x < 2_x
+        lt
+        # => [is_felt_lt, is_lhs_less, is_felt_eq, continue, <remaining_felts>]
+
+        dup.3 and
+        # => [is_felt_lt_if_continue, is_lhs_less, is_felt_eq, continue, <remaining_felts>]
+
+        or movdn.2
+        # => [is_felt_eq, continue, is_lhs_less, <remaining_felts>]
+
+        # keeps continue at 1 if the felts are equal
+        # sets continue to 0 if the felts are not equal
+        and
+        # => [continue, is_lhs_less, <remaining_felts>]
+
+        swap
+        # => [is_lhs_less, continue, <remaining_felts>]
+    end
+    # => [is_lhs_less, continue]
+
+    swap drop
+    # => [is_lhs_less]
+end
+
+#! Returns true if LHS is less than or equal to RHS.
+#!
+#! Inputs:  [LHS, RHS]
+#! Outputs: [is_lhs_less_or_equal]
+#!
+#! Cycles: 122
+export.lte
+    exec.gt
+    not
+end
+
+#! Returns true if LHS is exactly equal to RHS, false otherwise.
+#! The implementation does not branch, and always performs the same number of comparisons.
+#!
+#! Inputs:  [LHS, RHS]
+#! Outputs: [lhs_eq_rhs]
+#!
+#! Cycles: 26
+export.eq
```

Comment:
When would this be used instead of the `eqw` instruction?

Reply:
Possibly it may not be. But I got the impression that it *may* be desirable to remove the `eqw` instruction:

> word::eq and maybe even word::assert_eq which are now dedicated VM instructions.[^1]

I can update the docs to say that it is *currently* equivalent to the `eqw` instruction.

[^1]: https://github.com/0xMiden/miden-vm/issues/1928#issue-3187880636

Reply:
My original thinking was that it may be awkward to have some word-related functionality in stdlib and other in the instruction set (e.g., `std::word::lt` and `std::word::gt`, but `eqw`). But now that I think about it, we can't really make `assert_eqw` a procedure because we can specify error messages for procedures.

Regarding `eqw` - I don't have a strong preference: we could have it in both places, only as an instruction, or only as a library procedure.

Reply:
I would expect that as a library procedure only, it would be much more inefficient than as an instruction (currently 1 cycle). I don't know how much of a problem that is in practice, but to me that's the main reason to make something an instruction rather than a procedure (the other being assertions as you noted).